macros/core/externalData.pl

```diff
@@ -0,0 +1,123 @@
+################################################################################
+# WeBWorK Online Homework Delivery System
+# Copyright &copy; 2000-2022 The WeBWorK Project, https://github.com/openwebwork
+#
+# This program is free software; you can redistribute it and/or modify it under
+# the terms of either: (a) the GNU General Public License as published by the
+# Free Software Foundation; either version 2, or (at your option) any later
+# version, or (b) the "Artistic License" which comes with this package.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE.  See either the GNU General Public License or the
+# Artistic License for more details.
+################################################################################
+
+=head1 NAME
+
+externalData.pl - Macro that reads and writes data from a problem to the database.
+This is useful for using data common to multiple problems.
+
+=cut
+
+sub _init_externalData { }
+
+=head2 store_number_list
+
+This is a answer checker that checks only that there are a list of real number values
+and the purpose is to store in the database.
+
+=head3 Example
+
+Make a List object, but the contents doesn't matter.
+
+  $ans = List(0);
+
+Then in the visible problem block add the following
+
+  BEGIN_PGML
+  Enter a list of values as a vector.
+
+  [_____]{$ans->cmp(list_checker => store_number_list(), key => 'my_list')}
+  END_PGML
+
+Note that you will need to add a key (this should be unique for the problem set)
+
+=cut
+
+sub store_number_list {
+	return sub {
+		my ($correct, $student, $ansHash, $value) = @_;
+		my @errors;    # stores error messages
+
+		# Make sure that the key is set
+		my $key = $ansHash->{key};
+		warn 'You must pass in the key to save the data.' unless $key;
+
+		# Get the answer name for this input and store the name and key.
+		RECORD_EXTRA_ANSWERS("_ext_data:$ansHash->{ans_label}:numeric_list:$key");
```

Comment:
As per my comment on https://github.com/openwebwork/webwork2/pull/1879, remove the `:numeric_list:` type part of this.  That is not a good idea.  The same for `:string:` and `:real:` below.

Reply:
Also, the `RECORD_EXTRA_ANSWERS` call should not be in the checker.  That also should be before the checker in the `store_number_list` method.

Of course all of these changes need to be made to any other methods in this file.

Reply:
Ahh, I guess it needs to be here with this technique.  I just noticed you are obtaining the answer label from the answer hash.  Hmm.  This is probably not a good structure.  I don't think it is a good idea to call RECORD_EXTRA_ANSWERS in a checker.  That seems fragile.

Reply:
Do you have any suggestions to improve this?

Reply:
For now, just leave the `RECORD_EXTRA_ANSWERS` inside the checker.  It should be fine.  Just remove the data type specifier (`numeric_list`, `string`, `real`).  Trying to validate data types on the webwork side of things without the actual MathObject is unreliable at best.